### DISCUSSION TITLE
Why the IP keep changing after a restart of docker container?

### DISCUSSION THREAD

**Discussion by linkdesu** [omniedgeio/omniedge#44] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/858724/188812397-f94424c2-f5fc-4916-a284-8f3d22f76d97.png)

I tried to update my key of a instance runs in docker in my ubuntu machine today, after that I found two machine with same name appear in device list.

So I deteled one and then restart the container in docker, another machine with same name and different IP appears.

Is this only happens to instances running in docker? How can I keep the IP of an identical machine the same.

**Comment by yongqianme** [comment, marked as answer]
The problem is the **machineid**.  We use `github.com/panta/machineid ` in the omniedge cli to get the machine id from machine.
```
# https://github.com/omniedgeio/omniedge-cli/blob/8aaa5bfe1ac81e8ea135d7d025bdbc878a4466e8/utils.go
func RevealHardwareUUID() (string, error) {
	id, err := machineid.ID()
	if err != nil {
		return "", errors.New(fmt.Sprintf("Fail to generate hardware id, err is %+v", err))
	}
	id = strings.ToLower(strings.Replace(id, "-", "", -1))
	idBytes, err := hex.DecodeString(id)
	if err != nil {
		return "", errors.New(fmt.Sprintf("Fail to generate hardware id, err is %+v", err))
	}
	hardwareUUID, err := uuid.FromBytes(idBytes)
	if err != nil {
		return "", errors.New(fmt.Sprintf("Fail to generate hardware id, err is %+v", err))
	}
	return hardwareUUID.String(), nil
}
```

We got an error to get correct machine id from the docker or other system like openwrt. 

AT the moment we use `cp /proc/sys/kernel/random/uuid /etc/machine-id` for the temp solution before we find a final solution. 

```
#https://github.com/omniedgeio/omniedge-docker/blob/main/entrypoint

#!/bin/sh

set -e
cp /proc/sys/kernel/random/uuid /etc/machine-id

if [ ! -d /dev/net ]; then mkdir /dev/net; fi
if [ ! -e /dev/net/tun ]; then mknod /dev/net/tun c 10 200; fi

# echo 'net.ipv4.ip_forward = 1' | tee -a /etc/sysctl.conf
# sysctl -p /etc/sysctl.conf
login() {
  sleep 5
  omniedge login -s ${OMNIEDGE_SECURITYKEY}
}

connect() {
  sleep 5
  omniedge join -n ${OMNIEDGE_VIRUTALNETWORK_ID}
}

login && connect

```
For your case, the uuid will renew if you recreate a container.

**Reply by linkdesu** [reply]
Thanks for your help. 

I think maybe this can be a new argument when starting the cli or the container. Or, from my first impression, I think the secret key is the machine ID.

**Comment by codering** [comment]
I suggest adding an environment variable `OMNIEDGE_MACHINE_ID`, and then the entrypoint script should check for this variable first.

```sh
#!/bin/sh

set -e
if [ $OMNIEDGE_MACHINE_ID ];then
  echo $OMNIEDGE_MACHINE_ID > /etc/machine-id
else
  cp /proc/sys/kernel/random/uuid /etc/machine-id
fi
...
...
```

**Reply by Yong-OmniEdge** [reply]
Would you mind testing and submitting a PR ?

**Reply by codering** [reply]
I have submitted a PR

https://github.com/omniedgeio/omniedge-docker/pull/3
